src/moveit_studio_agent_examples/scripts/call_do_objective_waypoint.py

```diff
@@ -0,0 +1,157 @@
+
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import rclpy
+from rclpy.action import ActionClient
+from rclpy.node import Node
+import sys
+
+from moveit_msgs.msg import MoveItErrorCodes
+from moveit_studio_agent_msgs.action import DoObjectiveSequence
+from moveit_studio_behavior_msgs.msg import (
+    BehaviorParameter,
+    BehaviorParameterDescription,
+)
+
+
+class DoObjectiveSequenceClient(Node):
+    """
+    ROS 2 node that acts as an Action Client for the MoveIt Studio Agent's Objective Server.
+    Configured for the "Move to Joint State" waypoint, and adds a parameter override for waypoint names.
+    """
+
+    def __init__(self):
+        super().__init__("DoObjectiveSequence")
+        self._action_client = ActionClient(self, DoObjectiveSequence, "do_objective")
+
+    def send_goal(self, waypoint_name, cancel):
+        """
+        Sends a DoObjectiveSequence Goal to the Objective Server via the node's Action Client.
+
+        Args:
+            waypoint_name: the (string) name of a waypoint to move to.
+
+        Returns:
+            goal_future: a rclpy.task.Future to a rclpy.action.client.ClientGoalHandle.
+        """
+        goal_msg = DoObjectiveSequence.Goal()
+        goal_msg.objective_name = "Move to Joint State"
+        self.cancel = cancel
+
+        behavior_parameter = BehaviorParameter()
+        behavior_parameter.behavior_namespaces.append("move_to_joint_state")
+        behavior_parameter.description.name = "waypoint_name"
+        behavior_parameter.description.type = BehaviorParameterDescription.TYPE_STRING
+        behavior_parameter.string_value = waypoint_name
+        goal_msg.parameter_overrides = [behavior_parameter]
+
+        self._action_client.wait_for_server()
+        self._send_goal_future = self._action_client.send_goal_async(goal_msg)
+        self._send_goal_future.add_done_callback(self.goal_response_callback)
+        return self._send_goal_future
+
+    def goal_response_callback(self, future):
```

Comment:
Oof. Yes action servers have a lot of overhead but I think this is a great example of how to use their callbacks from a script.

Reply:
I agree, I find the callback part added is the part I (as a user) would actually want to see an example of how to do on the whole `Action Client` side of things. I debated passing the `cancel` argument to a callback function by passing a lambda where that's set (which appears to be the proper ros2 way of doing it) so they could see that, but figured for a quick example it may be simpler to just store it in the class and use `self.cancel` instead.